src/main/pacman/pacman_agent.py

```diff
@@ -0,0 +1,350 @@
+        model.compile(loss="mse", optimizer=Adam(lr=0.001), metrics=['accuracy'])
+        return model
+
+    def train_epochs(self, num_enemies, num_eps): # This function is only used when pacman_agent.py is run on its own
+        rewards_list = []
+        total_reward = 0
+        epoch_history = []
+        for ep in range(1, num_eps+1):
+            print("Running Episode No. ",ep)
+            env = PacmanEnv(model=self.model, num_enemies=num_enemies, ACTION_SPACE_SIZE=self.ACTION_SPACE_SIZE)
+
+            render = False
+            if ep%self.RENDER_EVERY == 0:
+                render = True
+
+            ep_history, _ = env.runEpisode(render)
+
+            [epoch_history.append(hist) for hist in ep_history]
+
+            if self.update_after == 'each':
+                self.update_model(epoch_history)
+            elif self.update_after != 'all':
+                print('update_after has to be either all or each')
+                assert False
+
+            # rewards_list.append(ep_reward)
+            # total_reward += ep_reward
+        if self.update_after=='all':
+            self.update_model(epoch_history)
+
+    def update_model(self, epoch_history): # Updates the model given the results of an epoch
+
+        if len(epoch_history)<self.BATCH_SIZE:
+            batch = epoch_history
+        else:
+            batch = random.sample(epoch_history, self.BATCH_SIZE)
+
+        # Get current states from minibatch, then query NN model for Q values
+        current_states = np.array([transition[0] for transition in batch]) / 255
+        current_qs_list = self.model.predict(current_states)
+
+        # Get future states from minibatch, then query NN model for Q values
+        # When using target network, query it, otherwise main network should be queried
+        new_current_states = np.array([transition[2] for transition in batch]) / 255
+        future_qs_list = self.model.predict(new_current_states)
+
+        X = []
+        y = []
+
+        # Now we need to enumerate our batches
+        for index, (current_state, action, new_current_state, reward, done) in enumerate(batch):
+
+            # If not a terminal state, get new q from future states, otherwise set it to 0
+            # almost like with Q Learning, but we use just part of equation here
+            if not done:
+                max_future_q = np.max(future_qs_list[index])
+                new_q = reward + self.DISCOUNT * max_future_q
+            else:
+                new_q = reward
+
+            # Update Q value for given state
+            current_qs = current_qs_list[index]
+            current_qs[action] = new_q
+
+            # And append to our training data
+            X.append(current_state)
+            y.append(current_qs)
+
+        # Fit on all samples as one batch, log only on terminal state
+        self.model.fit(np.array(X) / 255, np.array(y), batch_size=len(epoch_history), verbose=0, shuffle=False,
```

Comment:
do this division and casting to np.array outside fit; something like this 
X_train = np.array(X) / 255 ,  y_train = np.array(y)